Bulkloader (stage 1)

peterstace opened the pull request:
- Reads RDFs from file
- Converts to NQuads
- Create postings
- Writes postings to sorted and batched files
- All phase 2/3 code (merge sorted files and produce target badger instance) has been removed and can be reviewed later.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/1399)
<!-- Reviewable:end -->

manishrjain commented:
Reviewed 1 of 15 files at r1.
Review status: 1 of 15 files reviewed at latest revision, 25 unresolved discussions.

---

*[cmd/bulkloader/app.go, line 24 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtETbPUqsUXk5Z-LIFu:-KtETbPUqsUXk5Z-LIFv:bxfesrh) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L24)):*
> ```Go
> }
> 
> type app struct {
> ```

s/app/loader?

---

*[cmd/bulkloader/app.go, line 33 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtESw3GYP7jUohRLtzH:-KtESw3GYP7jUohRLtzI:bgkslh9) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L33)):*
> ```Go
> 
> func newApp(opt options) *app {
> ```

Extra vertical space.

---

*[cmd/bulkloader/app.go, line 75 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtET-VNjAD7tfbt-e21:-KtET-VNjAD7tfbt-e22:b-7u7nwn) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L75)):*
> ```Go
> 	defer f.Close()
> 
> 	var workerWG sync.WaitGroup
> ```

`Wg`.

---

*[cmd/bulkloader/progress.go, line 26 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtETreqFeXVfKVg8YYS:-KtETreqFeXVfKVg8YYT:b-ev00oa) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L26)):*
> ```Go
> }
> 
> func (p *progress) reportProgress() {
> ```

progress suffix is already present in the name. Remove it, so the api reads: `var pr progress; pr.report()`

---

*[cmd/bulkloader/progress.go, line 29 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEU1rLUQoBbpwhop87:-KtEU1rLUQoBbpwhop88:b-3k9maq) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L29)):*
> ```Go
> 	for {
> 		select {
> 		case <-time.After(time.Second):
> ```

Use a ticker. t := time.NewTicker(time.Second)

`for { select { case <- t.Ch: ...}}`

---

*[cmd/bulkloader/progress.go, line 32 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEUCY97w6JMHqVleHQ:-KtEUCY97w6JMHqVleHR:bdcq95k) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L32)):*
> ```Go
> 			p.report()
> 		case <-p.shutdown:
> 			p.shutdown <- struct{}{}
> ```

This pattern is strange (pulling from a stop chan, then pushing to the chan). You don't need this. You can instead just store the ticker variable, and call Stop on it from outside; which would exit this loop. So, this case isn't required.

---

*[cmd/bulkloader/progress.go, line 42 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEUd3ZumGBnBP-HA1u:-KtEUd3_0alCPl1f23d4:bvzxexo) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L42)):*
> ```Go
> 	elapsed := time.Since(p.start)
> 	fmt.Printf("[%s] [RDF count: %d] [RDFs per second: %d] [sorting: %d] [writing: %d]\n",
> 		round(elapsed).String(),
> ```

You can just do `duration.String()`, which would nicely format it for you.

What do sorting and writing do? Are they the different phases?

A typical mapreduce has 3 phases: Map, Shuffle, Reduce. You should use that terminology across this binary.

---

*[cmd/bulkloader/progress.go, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEV0YnHEeC2pp7r5__:-KtEV0YnHEeC2pp7r5_a:bxhrwb9) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L53)):*
> ```Go
> func (p *progress) endSummary() {
> 
> 	p.shutdown <- struct{}{}
> ```

ticker.Stop(). But then you need something to wait.

Actually, in Badger, I have a nice LevelCloser class. We could bring it into Dgraph x package, and use it for various purposes. It can be used to signal, then wait.

---

*[cmd/bulkloader/schema.go, line 18 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVYgIOEM5FUeo7fzb:-KtEVYgIOEM5FUeo7fzc:bjrnm32) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L18)):*
> ```Go
> 
> type schemaStore struct {
> 	mu sync.Mutex
> ```

There's only one variable here `m`. In this case, you can make sync.Mutex be the base class. Remove `mu` from front of sync.Mutex. Then you can call `s.Lock()` directly.

---

*[cmd/bulkloader/schema.go, line 41 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEY192XTeBsryblvZc:-KtEY192XTeBsryblvZd:bi5bvug) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L41)):*
> ```Go
> }
> 
> func (s *schemaStore) fixEdge(de *protos.DirectedEdge, objectIsUID bool) {
> ```

validateType instead of fixEdge. The former tells the reader the purpose of this function.

---

*[cmd/bulkloader/schema.go, line 42 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEYIDcAcG-ogmBG4jv:-KtEYIDcAcG-ogmBG4jw:b5ro2j5) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L42)):*
> ```Go
> 
> func (s *schemaStore) fixEdge(de *protos.DirectedEdge, objectIsUID bool) {
> 	s.mu.Lock()
> ```

I'd create a function part of schemaStore which can first acquire read lock, to check for the type. And only if it couldn't find it, then acquire a write lock. Note that after the first few thousand RDFs, we'd have seen most of the predicates, so this access would be reduced to read only. Using a write lock would be contentious.

---

*[cmd/bulkloader/schema.go, line 56 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEYEF1Png8t-68YRbg:-KtEYEF1Png8t-68YRbh:b-enr4bp) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L56)):*
> ```Go
> 
> 	schTyp := types.TypeID(sch.ValueType)
> 	err := wk.ValidateAndConvert(de, schTyp)
> ```

You don't need a lock around this particular logic, right?

---

*[cmd/bulkloader/worker.go, line 20 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVrt2utVpgw8lw-CT:-KtEVrt2utVpgw8lw-CU:b-ln41af) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L20)):*
> ```Go
> )
> 
> type worker struct {
> ```

mapper?

---

*[cmd/bulkloader/worker.go, line 25 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVz3gHrzwwEjDm5KV:-KtEVz3gHrzwwEjDm5KW:bh8tl4u) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L25)):*
> ```Go
> 	ss          *schemaStore
> 	prog        *progress
> 	postingsOut chan<- *protos.DenormalisedPosting
> ```

Denormalised is a long word. Consider switching it with `Flat`.

---

*[cmd/bulkloader/worker.go, line 28 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWHymImhHG59XDsvU:-KtEWHymImhHG59XDsvV:b-web46t) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L28)):*
> ```Go
> }
> 
> func newWorker(
> ```

This function is practically a what I call, creator function. It just creates a struct and give it back. We don't allow those in the code base. If you need to create a struct, do so directly without invoking a function.

---

*[cmd/bulkloader/worker.go, line 29 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEW5qGmA819j8me3jS:-KtEW5qHNpIN5OSTv3I0:b-u3ngge) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L29)):*
> ```Go
> 
> func newWorker(
> 	rdfCh chan string,
> ```

Don't put one arg per line. Fit as many as you can in one line, while making it look nice.

---

*[cmd/bulkloader/worker.go, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWdglZlxgBX3OGnWQ:-KtEWdglZlxgBX3OGnWR:b-rxbuvy) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L53)):*
> ```Go
> func (w *worker) addPosting(key []byte, posting *protos.Posting) {
> 	p := &protos.DenormalisedPosting{
> 		PostingListKey: key,
> ```

The struct name already has posting in there. So, no need to name internal variables `Posting`. Just call it Key.

---

*[cmd/bulkloader/worker.go, line 56 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWlKUL5RhHUGdF946:-KtEWlKUL5RhHUGdF947:bgk3lhs) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L56)):*
> ```Go
> 	}
> 	if posting.PostingType == protos.Posting_REF {
> 		p.Posting = &protos.DenormalisedPosting_UidPosting{UidPosting: posting.Uid}
> ```

The effect of long names can be seen here. They don't look good.

---

*[cmd/bulkloader/worker.go, line 63 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWyd-ApdRjSI3cr4b:-KtEWyd-ApdRjSI3cr4c:biof6jl) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L63)):*
> ```Go
> }
> 
> func (w *worker) parseRDF(rdfLine string) {
> ```

make this return error.

---

*[cmd/bulkloader/worker.go, line 69 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWxBoCZIL1p06fpcb:-KtEWxBoCZIL1p06fpcc:b-9966mg) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L69)):*
> ```Go
> 			return
> 		}
> 		x.Checkf(err, "Could not parse RDF.")
> ```

Don't check. Just return an error.

---

*[cmd/bulkloader/worker.go, line 72 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEX-Ohekh1F7igbdT6:-KtEX-Ohekh1F7igbdT7:b-sw9d97) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L72)):*
> ```Go
> 	}
> 
> 	sUID := w.um.assignUID(nq.GetSubject())
> ```

Don't worry about go linter asking you to have ID in caps. This is `sid`.

---

*[cmd/bulkloader/worker.go, line 73 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEX7Y5dQ_qRMNUkTkW:-KtEX7Y5dQ_qRMNUkTkX:b-1pf3d4) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L73)):*
> ```Go
> 
> 	sUID := w.um.assignUID(nq.GetSubject())
> 	uidM := map[string]uint64{nq.GetSubject(): sUID}
> ```

maps are expensive. Do you need it for parsing just one line?

---

*[cmd/bulkloader/worker.go, line 80 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXfIBpNZNHhddQHbA:-KtEXfIBpNZNHhddQHbB:b25e8zj) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L80)):*
> ```Go
> 	}
> 
> 	fwdPosting, revPosting := w.createEdgePostings(nq, uidM)
> ```

What is an edge Posting? Just call it Posting.

---

*[cmd/bulkloader/worker.go, line 93 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXP7mmEy9Uh16aFgL:-KtEXP7mmEy9Uh16aFgM:bcr7q8r) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L93)):*
> ```Go
> 
> 	w.addPosting(key, pp)
> ```

Remove vertical space.

---

*[cmd/bulkloader/worker.go, line 115 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXX6AnnmfbgAQdxn0:-KtEXX6AnnmfbgAQdxn1:b1cptx3) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L115)):*
> ```Go
> }
> 
> func (w *worker) createEdgePostings(nq gql.NQuad, uidM map[string]uint64) (*protos.Posting, *protos.Posting) {
> ```

Don't exceed 100 chars.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: 1 of 15 files reviewed at latest revision, 34 unresolved discussions.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEb-L8qlJv7z_W7e5_:-KtEb-L8qlJv7z_W7e5a:b-booxy6) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L1)):*
> ```Go
> package main
> ```

This file needs a shorter name. Just call it postings or something.

Also, there isn't as much logic here. Could be part of `mapper.go`.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 17 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbAwa-L6nIB2MSoxD:-KtEbAwa-L6nIB2MSoxE:boyr5aj) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L17)):*
> ```Go
> )
> 
> func writeDenormalisedPostings(dir string, postingsIn <-chan *protos.DenormalisedPosting, prog *progress) {
> ```

writePostings

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 17 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbZFeZDtGj_pDQNuU:-KtEbZFeZDtGj_pDQNuV:bxw72ya) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L17)):*
> ```Go
> )
> 
> func writeDenormalisedPostings(dir string, postingsIn <-chan *protos.DenormalisedPosting, prog *progress) {
> ```

Ch would be a more commonly used suffix.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 21 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbP9b-ulZLCT25tpM:-KtEbP9b-ulZLCT25tpN:bdxnxmf) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L21)):*
> ```Go
> 	var fileNum int
> 	var postings []*protos.DenormalisedPosting
> 	var wg sync.WaitGroup
> ```

Do a wg.Add(1) here as well; so we never reach zero, until just before wg.Wait().

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 26 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbmWco9VL-Pyvdr5d:-KtEbmWco9VL-Pyvdr5e:b-8te0t5) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L26)):*
> ```Go
> 	processBatch := func() {
> 		wg.Add(1)
> 		filename := filepath.Join(dir, fmt.Sprintf("%06d.bin", fileNum))
> ```

Did you ensure that the directory is empty before starting?

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 47 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbUR8QL-H62U2ynw7:-KtEbUR8QL-H62U2ynw8:b-fq9u9e) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L47)):*
> ```Go
> 		processBatch()
> 	}
> ```

wg.Done() here to tackle the Add at the top.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 51 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc96pP23GKkmaauVD:-KtEc96pP23GKkmaauVE:by6benz) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L51)):*
> ```Go
> }
> 
> func sortAndDump(filename string, postings []*protos.DenormalisedPosting, prog *progress) {
> ```

`sortAndWrite`

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 68 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc1_ME047xgnFGjDv:-KtEc1_ME047xgnFGjDw:bq13kjq) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L68)):*
> ```Go
> 	x.Checkf(err, "Could not open tmp file.")
> 	x.Check2(fd.Write(buf.Bytes()))
> 	x.Check(fd.Close())
> ```

fd.Sync before?

---

*[cmd/bulkloader/worker.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEcOdWq8eqB82XziYW:-KtEcOdWq8eqB82XziYX:b-2y3vnn) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L1)):*
> ```Go
> package main
> ```

Call it mapper.go

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 1 of 15 files reviewed at latest revision, 34 unresolved discussions.

---

*[cmd/bulkloader/app.go, line 24 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtETbPUqsUXk5Z-LIFu:-KtEaZQNwiyUKg5tXr8M:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L24)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

s/app/loader?
</blockquote></details>

Done.

---

*[cmd/bulkloader/app.go, line 33 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtESw3GYP7jUohRLtzH:-KtEaclh2t0XKCGV-gVS:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L33)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Extra vertical space.
</blockquote></details>

Done.

---

*[cmd/bulkloader/app.go, line 75 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtET-VNjAD7tfbt-e21:-KtEamiE7lpX7Bjenk39:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/app.go#L75)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

`Wg`.
</blockquote></details>

Done.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEb-L8qlJv7z_W7e5_:-KtFAmga6_gHcR2CAkRz:b-p2nk8f) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L1)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This file needs a shorter name. Just call it postings or something.

Also, there isn't as much logic here. Could be part of `mapper.go`.
</blockquote></details>

Renamed to `write_flat.go`.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 17 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbAwa-L6nIB2MSoxD:-KtEoU1ftVW9GOIkcZEd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L17)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

writePostings
</blockquote></details>

Done.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 17 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbZFeZDtGj_pDQNuU:-KtEoUmximcs-s6qxExr:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L17)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Ch would be a more commonly used suffix.
</blockquote></details>

Done.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 21 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbP9b-ulZLCT25tpM:-KtEp7G2QP7t6k7IryKN:b-exc1rl) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L21)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Do a wg.Add(1) here as well; so we never reach zero, until just before wg.Wait().
</blockquote></details>

I don't understand why we'd have to do this. I'll discuss in person.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 26 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbmWco9VL-Pyvdr5d:-KtEp4O7AI44ONABiYHc:b-jcvl86) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L26)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Did you ensure that the directory is empty before starting?
</blockquote></details>

It's created using `ioutil.TempDir`, so should be empty. The user can control the tmp dir prefix, so can control the device it gets created on, but not the absolute location. Is this what we want, or should be give control over the absolute location? (in which case, it would make 100% sense to ensure the location is empty).

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 51 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc96pP23GKkmaauVD:-KtEn3EBgPAY8CMNIIms:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L51)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

`sortAndWrite`
</blockquote></details>

Done.

---

*[cmd/bulkloader/denormalised_posting_writer.go, line 68 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc1_ME047xgnFGjDv:-KtEpnLtXWICzM733Euc:b-wr9dcf) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/denormalised_posting_writer.go#L68)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

fd.Sync before?
</blockquote></details>

I've added that now. I'm not sure what guarantees the file system gives for files that are unsynced but opened for reading later -- but if there's no guarantee that they'll be there, then the sync is important.

---

*[cmd/bulkloader/progress.go, line 26 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtETreqFeXVfKVg8YYS:-KtEbDKqUWt5AHDEAZVh:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L26)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

progress suffix is already present in the name. Remove it, so the api reads: `var pr progress; pr.report()`
</blockquote></details>

Done.

---

*[cmd/bulkloader/progress.go, line 42 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEUd3ZumGBnBP-HA1u:-KtEbmHZSdU3vtBC48wM:bh0qofs) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L42)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You can just do `duration.String()`, which would nicely format it for you.

What do sorting and writing do? Are they the different phases?

A typical mapreduce has 3 phases: Map, Shuffle, Reduce. You should use that terminology across this binary.
</blockquote></details>

**1.** `duration.String()` gives me:

```
[1.000167674s] [RDF count: 143703] [RDFs per second: 143678] [sorting: 0] [writing: 0]
[2.003535177s] [RDF count: 293721] [RDFs per second: 146601] [sorting: 0] [writing: 0]
[3.003660664s] [RDF count: 469184] [RDFs per second: 156204] [sorting: 0] [writing: 0]
[4.003778346s] [RDF count: 612598] [RDFs per second: 153004] [sorting: 0] [writing: 0]
[5.00459929s] [RDF count: 805075] [RDFs per second: 160867] [sorting: 0] [writing: 0]
[6.004721817s] [RDF count: 850424] [RDFs per second: 141625] [sorting: 1] [writing: 0]
[7.005076354s] [RDF count: 850424] [RDFs per second: 121401] [sorting: 1] [writing: 0]
[8.005152558s] [RDF count: 850424] [RDFs per second: 106234] [sorting: 1] [writing: 0]
[9.020774416s] [RDF count: 850424] [RDFs per second: 94273] [sorting: 0] [writing: 1]
[10.021070492s] [RDF count: 850424] [RDFs per second: 84863] [sorting: 0] [writing: 1]
[10.064847686s] [RDF count: 850424] [RDFs per second: 84494] [sorting: 0] [writing: 0]
```

which I think is a bit distracting to the user (all those decimal places). Rounding to a whole number of seconds seems more useful.

**2.** Sorting and writing are just counts of the current number of goroutines that are currently in the "sorting phase" and "writing phase". Useful for debugging, but not so much for the user. I've removed them.

**3.** Ack.

---

*[cmd/bulkloader/schema.go, line 18 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVYgIOEM5FUeo7fzb:-KtExjlKMzpjNKIRQypH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L18)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

There's only one variable here `m`. In this case, you can make sync.Mutex be the base class. Remove `mu` from front of sync.Mutex. Then you can call `s.Lock()` directly.
</blockquote></details>

Done.

---

*[cmd/bulkloader/schema.go, line 41 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEY192XTeBsryblvZc:-KtExzuMWCWMfijAvkFD:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L41)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

validateType instead of fixEdge. The former tells the reader the purpose of this function.
</blockquote></details>

Done.

---

*[cmd/bulkloader/schema.go, line 42 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEYIDcAcG-ogmBG4jv:-KtF0qq67rCzHAsT2kw_:b9z4c9d) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L42)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

I'd create a function part of schemaStore which can first acquire read lock, to check for the type. And only if it couldn't find it, then acquire a write lock. Note that after the first few thousand RDFs, we'd have seen most of the predicates, so this access would be reduced to read only. Using a write lock would be contentious.
</blockquote></details>

Good suggestion. I tried it out as a separate function, but since we need to keep track of how to unlock (via the `Unlock` or `RUnlock` method), it ended up being cleaner to just do it inline in the `validateType` function (that's the only place where we need a write lock).

---

*[cmd/bulkloader/schema.go, line 56 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEYEF1Png8t-68YRbg:-KtEySnRmd5DeaHX_Smm:b-4057yg) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/schema.go#L56)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

You don't need a lock around this particular logic, right?
</blockquote></details>

Correct, it's only needed for the map access. So not needed for the whole function. I've reduced the scope of where the lock is held to be more specific.

---

*[cmd/bulkloader/worker.go, line 1 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEcOdWq8eqB82XziYW:-KtF2M_4QWObEMwVx_3Z:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L1)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Call it mapper.go
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 20 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVrt2utVpgw8lw-CT:-KtF2N6ARhavj0cgEMqS:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L20)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

mapper?
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 25 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEVz3gHrzwwEjDm5KV:-KtF2W2HcSaAgKCFVEGp:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L25)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Denormalised is a long word. Consider switching it with `Flat`.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 28 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWHymImhHG59XDsvU:-KtF3kx25axYfi8DVDMc:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L28)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This function is practically a what I call, creator function. It just creates a struct and give it back. We don't allow those in the code base. If you need to create a struct, do so directly without invoking a function.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWdglZlxgBX3OGnWQ:-KtF2iOqNO-ZxcrwGOW6:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L53)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

The struct name already has posting in there. So, no need to name internal variables `Posting`. Just call it Key.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 63 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWyd-ApdRjSI3cr4b:-KtF4e4SDwnMbRtR5gt1:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L63)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

make this return error.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 69 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWxBoCZIL1p06fpcb:-KtF4e_4mjo211YZydXM:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L69)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't check. Just return an error.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 72 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEX-Ohekh1F7igbdT6:-KtF4qRxvGputkT8jYcG:bx6gx7x) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L72)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't worry about go linter asking you to have ID in caps. This is `sid`.
</blockquote></details>

Done. Same with `s/oUID/oid`.

---

*[cmd/bulkloader/worker.go, line 73 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEX7Y5dQ_qRMNUkTkW:-KtF4vZ1sFU0VYdVPD9G:bpn63fi) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L73)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

maps are expensive. Do you need it for parsing just one line?
</blockquote></details>

It's needed for the conversion function for NQuad to directed edge (in the `gql` package). I'll come back to this and add/modify to not require a map (can just pass the 2 IDs instead of the map).

---

*[cmd/bulkloader/worker.go, line 80 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXfIBpNZNHhddQHbA:-KtF5sr7FFJgWYpxIaEq:bx0z6z6) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L80)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

What is an edge Posting? Just call it Posting.
</blockquote></details>

I've referred to them as "edge postings" (i.e. posting for node to node or node to literal) to differentiate from "predicate posting" (i.e. posting from node to predicate name via the `"_predicate"` predicate) and "index postings" (i.e. posting from token/index-type to node).

On reflection, "edge" isn't a great name, since all postings are really edges. I've renamed to `createFwdAndRevPostings`, which I think better indicates what's going on here.

---

*[cmd/bulkloader/worker.go, line 93 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXP7mmEy9Uh16aFgL:-KtF7_-Pv4lpex8x_1ja:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L93)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove vertical space.
</blockquote></details>

Done.

---

*[cmd/bulkloader/worker.go, line 115 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEXX6AnnmfbgAQdxn0:-KtF86DjFp7KkZ5lCtzF:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/worker.go#L115)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't exceed 100 chars.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 0 of 16 files reviewed at latest revision, 34 unresolved discussions.

---

*[cmd/bulkloader/mapper.go, line 29 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEW5qGmA819j8me3jS:-KtFM1rAxh2jeWmNnnGI:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/mapper.go#L29)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't put one arg per line. Fit as many as you can in one line, while making it look nice.
</blockquote></details>

Done.

---

*[cmd/bulkloader/mapper.go, line 56 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEWlKUL5RhHUGdF946:-KtFMYP_HXofRymRCcdR:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/mapper.go#L56)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

The effect of long names can be seen here. They don't look good.
</blockquote></details>

Done.

---

*[cmd/bulkloader/mapper.go, line 73 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEX7Y5dQ_qRMNUkTkW:-KtFM_K9NY8gJZBfDo_q:b-7l9wxm) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/mapper.go#L73)):*
<details><summary><i>Previously, peterstace (Peter Stace) wrote…</i></summary><blockquote>

It's needed for the conversion function for NQuad to directed edge (in the `gql` package). I'll come back to this and add/modify to not require a map (can just pass the 2 IDs instead of the map).
</blockquote></details>

On followup for this, there are actually methods to do this without the map, I just had to export them. Done.

---

*[cmd/bulkloader/progress.go, line 53 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEV0YnHEeC2pp7r5__:-KtFR1F5seu0plHMWEBW:b-wp5l78) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L53)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

ticker.Stop(). But then you need something to wait.

Actually, in Badger, I have a nice LevelCloser class. We could bring it into Dgraph x package, and use it for various purposes. It can be used to signal, then wait.
</blockquote></details>

As we discussed in person, staying with the bidirectional channel to manage shutdown (using LevelCloser was not straight forward).

---

*[cmd/bulkloader/write_flat.go, line 47 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEbUR8QL-H62U2ynw7:-KtFQlvbivOwlz8hwl-X:b-n683gn) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/write_flat.go#L47)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

wg.Done() here to tackle the Add at the top.
</blockquote></details>

As discussed, don't need to do this (wg is still valid when it has a count of 0).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Comments.

---

Reviewed 7 of 18 files at r1, 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[cmd/bulkloader/loader.go, line 21 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFuYp1ztenB5bAZbh_:-KtFuYp1ztenB5bAZbha:b-3p0fc2) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L21)):*
> ```Go
> 	badgerDir  string
> 	tmpDir     string
> 	workers    int
> ```

mappers? Or, if this is for everything, then `numGoroutines`.

---

*[cmd/bulkloader/loader.go, line 26 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFxK_zJ9ZMGUO5xWa2:-KtFxK_zJ9ZMGUO5xWa3:bxsnzwl) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L26)):*
> ```Go
> type loader struct {
> 	opt        options
> 	prog       *progress
> ```

Make mapper, shuffle, reduce a subclass of a state that loader has.

---

*[cmd/bulkloader/loader.go, line 38 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtE4S15Os5oGvrzLq:-KtFtE4S15Os5oGvrzLr:b-xzvejv) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L38)):*
> ```Go
> 	x.Checkf(err, "Could not parse schema.")
> 
> 	a := &loader{
> ```

l := &loader

---

*[cmd/bulkloader/loader.go, line 41 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtGUqfDZ8v5DnDVZR:-KtFtGUqfDZ8v5DnDVZS:bhows4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L41)):*
> ```Go
> 		opt:        opt,
> 		prog:       newProgress(),
> 		rdfCh:      make(chan string, 1<<10),
> ```

1<<10 suggests size in bytes. But, this is just the number of strings. So, maybe use 1000.

---

*[cmd/bulkloader/loader.go, line 43 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtMWIF5vTZWLNzGdd:-KtFtMWIF5vTZWLNzGde:b-5h08c7) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L43)):*
> ```Go
> 		rdfCh:      make(chan string, 1<<10),
> 		mappers:    make([]*mapper, opt.workers),
> 		postingsCh: make(chan *protos.FlatPosting, 1<<10),
> ```

same here.

---

*[cmd/bulkloader/loader.go, line 51 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtSEsZUV0hNBOc3xN:-KtFtSEsZUV0hNBOc3xO:bu2br1b) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L51)):*
> ```Go
> 
> 	for i := 0; i < opt.workers; i++ {
> 		a.mappers[i] = &mapper{a.rdfCh, um, ss, a.prog, a.postingsCh}
> ```

When creating structs, also specify the keys.

---

*[cmd/bulkloader/loader.go, line 56 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtY-Bl29UVFgV6OEi:-KtFtY-Bl29UVFgV6OEj:b-hgfdjw) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L56)):*
> ```Go
> }
> 
> func (a *loader) run() {
> ```

l *loader

---

*[cmd/bulkloader/loader.go, line 57 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtVyomZHx3KYBqljp:-KtFtVyomZHx3KYBqljq:bgg9ymq) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L57)):*
> ```Go
> 
> func (a *loader) run() {
> ```

vert space.

---

*[cmd/bulkloader/loader.go, line 61 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtg0mib1LPRBwImuw:-KtFtg0mib1LPRBwImux:b-d5sgoy) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L61)):*
> ```Go
> 
> 	var postingWriterWg sync.WaitGroup
> 	postingWriterWg.Add(1)
> ```

Space after this?

---

*[cmd/bulkloader/loader.go, line 64 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtf140cD7mRaJzSo2:-KtFtf140cD7mRaJzSo3:b-d5sgoy) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L64)):*
> ```Go
> 	tmpPostingsDir, err := ioutil.TempDir(a.opt.tmpDir, "bulkloader_tmp_posting_")
> 	x.Check(err)
> 	defer func() { x.Check(os.RemoveAll(tmpPostingsDir)) }()
> ```

Space after this?

---

*[cmd/bulkloader/loader.go, line 77 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtpetJPtsXoj2_utM:-KtFtpetJPtsXoj2_utN:bb54eo5) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L77)):*
> ```Go
> 	mapperWg.Add(len(a.mappers))
> 	for _, m := range a.mappers {
> 		m := m
> ```

Ahh.. interesting. But, it's more idiomatic to do `go func(m *mapper) {}(m)`.

---

*[cmd/bulkloader/mapper.go, line 65 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwMARR1tCGJDwkhFt:-KtFwMARR1tCGJDwkhFu:b-citnke) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L65)):*
> ```Go
> 	}
> 
> 	fwdPosting, revPosting := m.createFwdAndRevPostings(nq, de)
> ```

Maybe just call it `createPostings`.

`fwd, rev := createPostings(...)`

---

*[cmd/bulkloader/mapper.go, line 132 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwHa2dbTceUJtWz0c:-KtFwHa2dbTceUJtWz0d:b-7jbqan) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L132)):*
> ```Go
> 
> func (m *mapper) addIndexPostings(nq gql.NQuad, de *protos.DirectedEdge) {
> ```

space

---

*[cmd/bulkloader/mapper.go, line 155 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwfBkpPB-bAUfXvdQ:-KtFwfBkpPB-bAUfXvdR:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L155)):*
> ```Go
> 		// Convert from storage type to schema type.
> 		schemaVal, err := types.Convert(storageVal, types.TypeID(sch.GetValueType()))
> 		x.Check(err) // Shouldn't error, since we've already checked for convertibility when doing edge postings.
> ```

100 chars.

---

*[cmd/bulkloader/progress.go, line 32 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEUCY97w6JMHqVleHQ:-KtFyEAZnIHekR9MgZRC:b2nhktp) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/progress.go#L32)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This pattern is strange (pulling from a stop chan, then pushing to the chan). You don't need this. You can instead just store the ticker variable, and call Stop on it from outside; which would exit this loop. So, this case isn't required.
</blockquote></details>

OK.

---

*[cmd/bulkloader/progress.go, line 14 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtG-HQn-Vj7LRF1uAD2:-KtG-HQn-Vj7LRF1uAD3:bmz6y08) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/progress.go#L14)):*
> ```Go
> 	start        time.Time
> 
> 	// shotdown is a bidirectional channel used to manage the stopping of the
> ```

s/shotdown/shutdown

---

*[cmd/bulkloader/schema.go, line 56 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFzAEBH8JnbTB1XEr6:-KtFzAEBH8JnbTB1XEr7:boku84y) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/schema.go#L56)):*
> ```Go
> 			sch = schemaState{false, &protos.SchemaUpdate{ValueType: de.ValueType}}
> 			s.m[de.Attr] = sch
> 		}
> ```

s.Unlock() here. No need for write variable.

---

*[cmd/bulkloader/schema.go, line 57 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFz8FCmmTXBvAdJDGw:-KtFz8FCmmTXBvAdJDGx:bb6vvo2) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/schema.go#L57)):*
> ```Go
> 			s.m[de.Attr] = sch
> 		}
> 	}
> ```

else { s.RUnlock() }

---

*[cmd/bulkloader/uid_map.go, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFzKv8Ba_GN-89lO8Y:-KtFzKv8Ba_GN-89lO8Z:b-jlj70j) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/uid_map.go#L12)):*
> ```Go
> 
> type uidMap struct {
> 	mu      sync.Mutex
> ```

Don't need mu here as well. Make it subclass of sync.Mutex.

---

*[cmd/bulkloader/write_flat.go, line 68 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc1_ME047xgnFGjDv:-KtFzkmPHGDep6HiW7PT:b-9hg1hu) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/write_flat.go#L68)):*
<details><summary><i>Previously, peterstace (Peter Stace) wrote…</i></summary><blockquote>

I've added that now. I'm not sure what guarantees the file system gives for files that are unsynced but opened for reading later -- but if there's no guarantee that they'll be there, then the sync is important.
</blockquote></details>

Hmm.. we probably would want a way to resume from a successful map phase. So, having all files synced before map completes is good.

Btw, might want to have a STATUS file, which tracks which phase we're at. So, resuming is easy.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-:-KtG-PY4UJLsOVRmHgDT:b9wrqbv)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[cmd/bulkloader/loader.go, line 21 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFuYp1ztenB5bAZbh_:-KtG2ZEXlL-qBIRBHF92:bbxybt2) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L21)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

mappers? Or, if this is for everything, then `numGoroutines`.
</blockquote></details>

Done, `numGoroutines`.

---

*[cmd/bulkloader/loader.go, line 26 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFxK_zJ9ZMGUO5xWa2:-KtG5Kz6rFf2UPBEzMuO:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L26)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Make mapper, shuffle, reduce a subclass of a state that loader has.
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 38 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtE4S15Os5oGvrzLq:-KtG5VRQvN34HK2RTRie:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L38)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

l := &loader
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 41 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtGUqfDZ8v5DnDVZR:-KtG5ar4emO4LiKssR4E:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L41)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

1<<10 suggests size in bytes. But, this is just the number of strings. So, maybe use 1000.
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 43 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtMWIF5vTZWLNzGdd:-KtG5bdDL6il2P_Et-Qw:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L43)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

same here.
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 51 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtSEsZUV0hNBOc3xN:-KtG5hhvxt8xwrG5SJzA:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L51)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

When creating structs, also specify the keys.
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 56 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtY-Bl29UVFgV6OEi:-KtG68NfgEjNFz1jDLCI:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L56)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

l *loader
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 57 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtVyomZHx3KYBqljp:-KtG67xh8GJfmYoAjPwR:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L57)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

vert space.
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 61 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtg0mib1LPRBwImuw:-KtG6IUvsc54jUQL6xNd:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L61)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Space after this?
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 64 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtf140cD7mRaJzSo2:-KtG6ItAd3RCT7s5JwE4:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L64)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Space after this?
</blockquote></details>

Done.

---

*[cmd/bulkloader/loader.go, line 77 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFtpetJPtsXoj2_utM:-KtG6U5wZwjvEoKZ5_4z:b2li1q9) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/loader.go#L77)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Ahh.. interesting. But, it's more idiomatic to do `go func(m *mapper) {}(m)`.
</blockquote></details>

I've seen it done both ways, happy to use a function param instead.

---

*[cmd/bulkloader/mapper.go, line 132 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwHa2dbTceUJtWz0c:-KtG6ndPLcJ4yNSnHHg8:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L132)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

space
</blockquote></details>

Done.

---

*[cmd/bulkloader/mapper.go, line 155 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwfBkpPB-bAUfXvdQ:-KtG714RzldZOLbVkB4U:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L155)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

100 chars.
</blockquote></details>

Done.

---

*[cmd/bulkloader/progress.go, line 14 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtG-HQn-Vj7LRF1uAD2:-KtG74_EY92nviIBPM97:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/progress.go#L14)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

s/shotdown/shutdown
</blockquote></details>

Done.

---

*[cmd/bulkloader/schema.go, line 56 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFzAEBH8JnbTB1XEr6:-KtG8fwor-0hNmHt_mt6:bh6wuki) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/schema.go#L56)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

s.Unlock() here. No need for write variable.
</blockquote></details>

Ah, yes, there's a simpler way to do this. I followed your suggestion, and realized I could go even simpler (removed the else clause completely as well).

---

*[cmd/bulkloader/schema.go, line 57 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFz8FCmmTXBvAdJDGw:-KtG8few4Zwqz9nTZviH:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/schema.go#L57)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

else { s.RUnlock() }
</blockquote></details>

Done.

---

*[cmd/bulkloader/uid_map.go, line 12 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFzKv8Ba_GN-89lO8Y:-KtG9ALQ0g3MKSlxky_g:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/uid_map.go#L12)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Don't need mu here as well. Make it subclass of sync.Mutex.
</blockquote></details>

Done.

---

*[cmd/bulkloader/write_flat.go, line 68 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtEc1_ME047xgnFGjDv:-KtG9R5CdllE28M3Nizn:bomqknw) ([raw file](https://github.com/dgraph-io/dgraph/blob/9d535aad35c3ffe006740f35622d90c1e45720b0/cmd/bulkloader/write_flat.go#L68)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Hmm.. we probably would want a way to resume from a successful map phase. So, having all files synced before map completes is good.

Btw, might want to have a STATUS file, which tracks which phase we're at. So, resuming is easy.
</blockquote></details>

Good suggestion. I'll add that as a TODO for once we have everything up and running end to end.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 10 of 16 files reviewed at latest revision, 21 unresolved discussions.

---

*[cmd/bulkloader/mapper.go, line 65 at r2](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399#-KtFwMARR1tCGJDwkhFt:-KtGAy160lBnDRQVeys9:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/f2688bf55981b0243952d1a5c57c8a777b80d977/cmd/bulkloader/mapper.go#L65)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Maybe just call it `createPostings`.

`fwd, rev := createPostings(...)`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/1399)*
<!-- Sent from Reviewable.io -->
